### PR TITLE
DOCSP-16500-add-snippets-feature

### PR DESCRIPTION
**Pull request by davemungo**
JIRA: 
https://jira.mongodb.org/browse/DOCSP-16500

STAGING:
https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/

https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/working-with-snippets/

https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/packages/

https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/registries-config/

https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/commands/

https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/configuration/

https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/registry-index/

https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/error-handlers/

https://docs-mongodbcom-staging.corp.mongodb.com/mongodb-shell/docsworker-xlarge/DOCSP-16500-add-snippets-feature/snippets/troubleshooting/

### CODE REVIEW

File: source/snippets/configuration.txt [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+Add a Second Registry
+~~~~~~~~~~~~~~~~~~~~~
+
+Configure a second, private registry for sensitive snippets by adding a
+URL to ``snippetIndexSourceURLs``. 
+
+.. code-block:: shell
+
+   config.set('snippetIndexSourceURLs', 
+     'https://github.com/YOUR_COMPANY/PATH_TO_YOUR_REGISTRY/index.bson.br;https://compass.mongodb.com/mongosh/snippets-index.bson.br'
```
Comment (addaleax) [inline source/snippets/configuration.txt:84]: (same as above:)

```suggestion
     'https://github.com/YOUR_COMPANY/PATH_TO_YOUR_REGISTRY/index.bson.br;' + config.get('snippetIndexSourceURLs')
```

File: source/snippets/packages.txt [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+The parameters are:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 33 67
+
+   * - Field
+     - Description
+   * - "name"
+     - The package that contains the snippet.
```
Comment (addaleax) [inline source/snippets/packages.txt:86]:
```suggestion
     - The npm package that contains the snippet.
```

File: source/snippets/packages.txt [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+       `SPDX license list <https://www.npmjs.com/package/spdx>`__. See
+       also the
+       `MongoDB Contributor Agreement <https://www.mongodb.com/legal/contributor-agreement>`__.
+   * - "publishConfig"
+     - This value is used to control access to your snippet package.
+       ``public`` is typical, but npm provides
+       `other options <https://docs.npmjs.com/cli/v7/using-npm/config>`__
+       as well.
+
+Use this code to create a skelton ``package.json`` file. Edit the file
```
Comment (addaleax) [inline source/snippets/packages.txt:113]:
```suggestion
Use this code to create a skeleton ``package.json`` file. Edit the file
```

File: source/snippets/registries-config.txt [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+to an index file that contains metadata for the snippets in that
+registry. 
+
+Configure an additional registry by adding a URL to
+``snippetIndexSourceURLs``. 
+
+.. code-block:: shell
+
+   config.set('snippetIndexSourceURLs', 
+     'https://github.com/YOUR_COMPANY/PATH_TO_YOUR_REPOSITORY/index.bson.br;https://compass.mongodb.com/mongosh/snippets-index.bson.br'
```
Comment (addaleax) [inline source/snippets/registries-config.txt:110]: (and one more:)

```suggestion
     'https://github.com/YOUR_COMPANY/PATH_TO_YOUR_REPOSITORY/index.bson.br;' + config.get('snippetIndexSourceURLs')
```

File: source/snippets/registry-index.txt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+   .. include:: /includes/warnings/warn-experimental-feature.rst
+
+This page discusses how to manually create a registry index file. To
+generate the registry index file using a script, see :ref:`Create a
+Registry Index File <snip-make-a-registry-index>`. 
+
+To create the registry index file:
+
+#. Copy the following TypeScript template and save it as ``index.bson``:
```
Comment (addaleax) [inline source/snippets/registry-index.txt:26]: Uh, so … this is the typescript *definitions* for what `index.bson` contains, but it’s not the actual contents. `index.bson` is actual BSON, i.e. binary data, not text.

I think that as this is for fairly advanced users only, it’s okay to leave out things like how to come up with the actual contents of the file (e.g. in the case of the default repository, the information is grabbed from the individual snippets’ package.json files).

I’d suggest adding 

```js
import bson from 'bson';
import zlib from 'zlib';
``` 

lines at the top of the typescript example here, and then adding the following at the bottom:

```js
const indexFileContents: SnippetIndexFile = {
   indexFileVersion: 1,
   index: [ /* ... */ ],
   metadata: { homepage: 'https://example.com' }
};

// Serialize, compress and store the index file:
fs.writeFileSync('index.bson.br',
   zlib.brotliCompressSync(
      bson.serialize(indexFileContents)));
```

The file can then be store as e.g. `make-index.ts`, and can be run as `ts-node make-index.ts` in order to actually create the `index.bson.br` file.

File: source/snippets/troubleshooting.txt [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+   /<NPM USER HOME>/.npm/_logs/2021-09-16T22_03_34_534Z-debug.log
+
+When you locate the log files, check the most recent one.
+
+Non-specific Error Message
+--------------------------
+
+**Problem**: :binary:`~bin.mongosh` returns a non-specific error
+message when you try to start the shell.
```
Comment (addaleax) [inline source/snippets/troubleshooting.txt:40]: I assume this is something that you ran into at some point – do you remember details? The right solution here is to actually provide a specific error message, of course :)

File: source/write-scripts/require-external-modules.txt [hunk 7]
```diff
@@ -108,6 +114,19 @@ install the modules either:
 - Globally
 - In the ``node_modules`` directory in your current working directory.
 
+.. tip::
+
+   If you package your script as a :ref:`snippet <snip-overview>`, you
```
Comment (addaleax) [inline source/write-scripts/require-external-modules.txt:119]: This is not just the case for snippets, but for all mongosh scripts.

**Review by addaleax — COMMENTED** [review]
Nice work! It will be really nice to have something to point users to when they ask about this :)

File: source/includes/steps-define-a-new-registry.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+content: |
+
+   To make the new registry available to your local
+   :binary:`~bin.mongosh` installation, update the
+   ``snippetIndexSourceURLs`` configuration settings.
+
+   .. code-block::
+
+      config.set('snippetIndexSourceURLs', 
+      'https://github.com/YOUR_COMPANY/PATH_TO_YOUR_REPOSITORY/index.bson.br;https://compass.mongodb.com/mongosh/snippets-index.bson.br'
```
[the remark this answers is not in the document]
Comment (davemungo) [reply source/includes/steps-define-a-new-registry.yaml:41]: Done

File: source/reference/configure-shell-settings.txt [hunk 9]
```diff
@@ -111,6 +113,25 @@ Supported ``property`` parameters
...
+     - true
+     - If ``true``, automatically load installed
+       :ref:`snippets <snip-overview>` at startup.
+
+   * - ``snippetIndexSourceURLs``
+     - string
+     - `MongoDB Repository
+       <https://compass.mongodb.com/mongosh/snippets-index.bson.br>`__
+     - A URL that links to a :ref:`snippet <snip-overview>`
+       registry.
```
[the remark this answers is not in the document]
Comment (davemungo) [reply source/reference/configure-shell-settings.txt:127]: Done

File: source/snippets/configuration.txt (hunk 1 above)
Comment (davemungo) [reply source/snippets/configuration.txt:84]: Done

File: source/snippets/packages.txt (hunk 2 above)
Comment (davemungo) [reply source/snippets/packages.txt:86]: Done

File: source/snippets/packages.txt (hunk 3 above)
Comment (davemungo) [reply source/snippets/packages.txt:113]: Done

File: source/snippets/registries-config.txt (hunk 4 above)
Comment (davemungo) [reply source/snippets/registries-config.txt:110]: Done

File: source/snippets/troubleshooting.txt (hunk 6 above)
Comment (davemungo) [reply source/snippets/troubleshooting.txt:40]: It was, I'm not sure what though. Most likely it happened when I was testing out config settings. I'll let you know if I run into it again.

File: source/write-scripts/require-external-modules.txt (hunk 7 above)
Comment (davemungo) [reply source/write-scripts/require-external-modules.txt:119]: Done

File: source/snippets/registry-index.txt (hunk 5 above)
Comment (davemungo) [reply source/snippets/registry-index.txt:26]: Done

**Review by addaleax — APPROVED** [review]

**Comment by davemungo** [comment]
Thanks Anna, I think this will be a popular feature for the shell.

**Comment by jeff-allen-mongo** [comment]
Hey @davemungo , this PR modified the .gitignore and .DS_store files. Was that intentional?

**Pull request merged**